File: web/src/lib/link.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

/**
 * Download links, per spec §10.
 *
 * The secret lives in the fragment, which browsers do not transmit. That is
 * what keeps it out of server logs, proxy logs and CDN access logs, and it is
 * also what makes a truncated link silently useless: the part most likely to be
 * lost in copying is the only part that cannot be recovered.
 */

import { FILE_ID_SIZE, LINK_SECRET_SIZE } from "../crypto/index.js";

/** Unpadded base64url, the wire format's encoding for binary values (spec §1). */
export function toBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i += 0x8000) {
    binary += String.fromCharCode(...bytes.subarray(i, i + 0x8000));
  }
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replaceAll("=", "");
}

/** The inverse of {@link toBase64Url}. Returns null for anything malformed. */
export function fromBase64Url(text: string): Uint8Array | null {
  // Checked rather than left to atob, which accepts the standard alphabet and
  // ignores some malformed input. A link is attacker-supplied text.
  if (text === "" || !/^[A-Za-z0-9_-]+$/.test(text)) return null;

  const padded = text
    .replaceAll("-", "+")
    .replaceAll("_", "/")
    .padEnd(text.length + ((4 - (text.length % 4)) % 4), "=");
  try {
    const binary = atob(padded);
    return Uint8Array.from(binary, (c) => c.charCodeAt(0));
  } catch {
    return null;
  }
}

/**
 * Builds the link a recipient opens.
 *
 * `origin` is passed in rather than read from `location` so that the value is
 * visible to the caller that displays it: a link built from an origin nobody
 * chose is how a client ends up handing out `http://localhost` links.
 */
export function downloadLink(origin: string, fileID: Uint8Array, linkSecret: Uint8Array): string {
  if (fileID.length !== FILE_ID_SIZE) {
    throw new TypeError(`link: file id is ${fileID.length} bytes, want ${FILE_ID_SIZE}`);
  }
  if (linkSecret.length !== LINK_SECRET_SIZE) {
    throw new TypeError(
      `link: link secret is ${linkSecret.length} bytes, want ${LINK_SECRET_SIZE}`,
    );
  }
  const base = origin.replace(/\/+$/, "");
  return `${base}/d/${toBase64Url(fileID)}#${toBase64Url(linkSecret)}`;
}

/** What a link resolves to. */
export interface ParsedLink {
  fileID: Uint8Array;
  linkSecret: Uint8Array;
}

/**
 * Recovers the identifier and secret from a link.
 *
 * Sizes are checked, so a link that lost characters in transit is reported as
 * broken rather than carried forward into a derivation that produces keys which
 * merely fail to open anything. Which of the two happened is the difference
 * between "this link is damaged" and "this file is corrupt".
 */
export function parseLink(link: string): ParsedLink | null {
  let url: URL;
  try {
    url = new URL(link);
  } catch {
    return null;
  }

  const match = /^\/d\/([A-Za-z0-9_-]+)$/.exec(url.pathname);
  if (match === null) return null;

  // A fragment that is present but empty is a link that was copied up to and
  // including the separator, which is exactly the failure this guards.
  const fragment = url.hash.startsWith("#") ? url.hash.slice(1) : "";

  const fileID = fromBase64Url(match[1] as string);
  const linkSecret = fromBase64Url(fragment);
  if (fileID === null || linkSecret === null) return null;
  if (fileID.length !== FILE_ID_SIZE || linkSecret.length !== LINK_SECRET_SIZE) return null;

  return { fileID, linkSecret };
}

/**
 * Whether the link is complete.
 *
 * Offered separately so an interface can say *what* is wrong. A link whose
 * fragment was dropped parses as far as the identifier, and telling the person
 * holding it that the secret is missing is actionable; "invalid link" is not.
 */
export function fragmentIsPresent(link: string): boolean {
  const hash = link.indexOf("#");
  return hash !== -1 && hash !== link.length - 1;
}

File: internal/client/link.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

// Package client speaks the Sendan API from the outside.
//
// It is the Go half of what web/src/lib does in the browser, and exists so the
// command line client and the browser are two front ends over one
// implementation of the scheme rather than two implementations of it. The
// cryptography itself is internal/crypto, shared with the server.
package client

import (
	"encoding/base64"
	"fmt"
	"net/url"
	"regexp"
	"strings"

	"github.com/Serraniel/sendan/internal/crypto"
)

// Link is what a recipient is given: an upload to fetch and the secret that
// opens it.
type Link struct {
	// Origin is the instance, without a trailing slash.
	Origin     string
	FileID     []byte
	LinkSecret []byte
}

// encoding is unpadded base64url, the wire format's spelling for binary values
// (spec §1).
var encoding = base64.RawURLEncoding

var idPattern = regexp.MustCompile(`^/d/([A-Za-z0-9_-]+)$`)

// String renders the link a recipient opens, per spec §10.
//
// The secret goes in the fragment, which browsers do not transmit. That is what
// keeps it out of server, proxy and CDN logs, and it is also why a link that
// loses its tail is silently useless: the part most likely to be dropped in
// copying is the only part that cannot be recovered.
func (l Link) String() string {
	return fmt.Sprintf("%s/d/%s#%s",
		strings.TrimRight(l.Origin, "/"),
		encoding.EncodeToString(l.FileID),
		encoding.EncodeToString(l.LinkSecret))
}

// ID is the upload's identifier as it appears in a request path.
func (l Link) ID() string { return encoding.EncodeToString(l.FileID) }

// ParseLink reads a link a person supplied.
//
// The sizes are checked, so a link that lost characters in transit is reported
// as damaged rather than carried into a derivation that produces keys which
// merely fail to open anything. Which of the two happened is the difference
// between "this link is broken" and "this file is corrupt", and only one of
// them is worth asking the sender about.
func ParseLink(raw string) (Link, error) {
	u, err := url.Parse(strings.TrimSpace(raw))
	if err != nil {
		return Link{}, fmt.Errorf("client: not a link: %w", err)
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return Link{}, fmt.Errorf("client: %q is not an http or https link", raw)
	}

	match := idPattern.FindStringSubmatch(u.Path)
	if match == nil {
		return Link{}, fmt.Errorf("client: %q does not address an upload", u.Path)
	}

	fileID, err := encoding.DecodeString(match[1])
	if err != nil || len(fileID) != crypto.FileIDSize {
		return Link{}, fmt.Errorf("client: the upload identifier in this link is damaged")
	}

	if u.Fragment == "" {
		return Link{}, fmt.Errorf(
			"client: this link is missing the part after the #, which is the key. " +
				"It cannot be recovered - ask the sender for the whole link")
	}
	secret, err := encoding.DecodeString(u.Fragment)
	if err != nil || len(secret) != crypto.LinkSecretSize {
		return Link{}, fmt.Errorf("client: the key in this link is damaged or incomplete")
	}

	origin := u.Scheme + "://" + u.Host
	return Link{Origin: origin, FileID: fileID, LinkSecret: secret}, nil
}

// EncodeToken renders a token for a person to keep, in the wire format's
// encoding so it can be pasted back wherever one is accepted.
func EncodeToken(token []byte) string { return encoding.EncodeToString(token) }
